// @moovio/node/dist/cjs/enrichedProfile.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnrichedProfiles = void 0;
const checks_js_1 = require("./helpers/checks.js");
const errors_js_1 = require("./helpers/errors.js");
/**
 * @typedef EnrichedProfile
 * @property {EnrichedIndividualProfile} individual - Describes a person
 * @property {EnrichedBusinessProfile} business - Describes a company
 *
 * @tag Enrichment
 */
/**
 * @typedef EnrichedBusinessProfile
 * @property {string} legalBusinessName
 * @property {EnrichedProfileAddress} address
 * @property {string} email
 * @property {EnrichedProfilePhone} phone
 * @property {EnrichedProfileIndustry} industryCodes - Describes industry specific identifiers
 * @property {string} website
 *
 * @tag Enrichment
 */
/**
 * @typedef EnrichedIndividualProfile
 * @property {EnrichedProfileName} name
 * @property {string} email
 * @property {EnrichedProfileAddress} address
 *
 * @tag Enrichment
 */
/**
 * @typedef EnrichedProfileAddress
 * @property {string} addressLine1
 * @property {string} addressLine2
 * @property {string} city
 * @property {string} stateOrProvince
 * @property {string} postalCode
 * @property {string} country
 *
 * @tag Enrichment
 */
/**
 * @typedef EnrichedProfileIndustry - Describes industry specific identifiers
 * @property {string} naics
 * @property {string} sic
 *
 * @tag Enrichment
 */
/**
 * @typedef EnrichedProfileName
 * @property {string} firstName
 * @property {string} middleName
 * @property {string} lastName
 * @property {string} suffix
 *
 * @tag Enrichment
 */
/**
 * @typedef EnrichedProfilePhone
 * @property {string} number
 * @property {string} countryCode
 *
 * @tag Enrichment
 */
/**
 * The Enriched Profile API.
 * @tag Enrichment
 */
class EnrichedProfiles {
    constructor(moov) {
        /**
         * @type {Moov}
         * @private
         */
        this.moov = moov;
    }
    /**
     * Gets enriched profile data.
     *
     * @param {string} email - Email address associated with the profile.
     * @returns {Promise<EnrichedProfile>}
     * @tag Enrichment
     *
     * @example
     * const moov = new Moov(...);
     * try {
     *   const enrichedProfile = moov.enrichedProfiles.get("employee@business.com");
     * } catch (err) {
     *   // ..
     * }
     */
    get(email) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, checks_js_1.checkString)(email).or(errors_js_1.Err.MISSING_EMAIL);
            const result = yield this.moov
                .got({
                url: "enrichment/profile",
                method: "GET",
                searchParams: {
                    email: email,
                },
            })
                .json();
            return result;
        });
    }
}
exports.EnrichedProfiles = EnrichedProfiles;
